abdm_integrator/user_auth/views.py

```diff
@@ -0,0 +1,133 @@
+        if response_data and response_data.get('auth'):
+            if AuthenticationMode.DIRECT in response_data['auth']['modes']:
+                response_data['auth']['modes'].remove(AuthenticationMode.DIRECT)
+            # Invalid Password mode is being returned.
+            # See https://devforum.abdm.gov.in/t/care-context-link-otp-not-received/8006/2?u=ayadav
+            if AuthenticationMode.PASSWORD in response_data['auth']['modes']:
+                response_data['auth']['modes'].remove(AuthenticationMode.PASSWORD)
+        return response_data
+
+
+class GatewayAuthOnFetchModes(UserAuthGatewayBaseView):
+
+    def post(self, request, format=None):
+        serializer = GatewayAuthOnFetchModesSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        ABDMCache.set(serializer.data['resp']['requestId'], serializer.data, CALLBACK_RESPONSE_CACHE_TIMEOUT)
+        return Response(status=HTTP_202_ACCEPTED)
+
+
+class AuthInit(UserAuthBaseView):
+
+    def post(self, request, format=None):
+        serializer = AuthInitSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        gateway_request_id = self.gateway_auth_init(serializer.data)
+        response_data = poll_for_data_in_cache(gateway_request_id)
+        return self.generate_response_from_callback(response_data)
+
+    def gateway_auth_init(self, request_data):
+        payload = ABDMRequestHelper.common_request_data()
+        payload['query'] = request_data
+        ABDMRequestHelper().gateway_post(UserAuthGatewayAPIPath.AUTH_INIT, payload)
+        return payload['requestId']
+
+
+class GatewayAuthOnInit(UserAuthGatewayBaseView):
+
+    def post(self, request, format=None):
+        serializer = GatewayAuthOnInitSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        ABDMCache.set(serializer.data['resp']['requestId'], serializer.data, CALLBACK_RESPONSE_CACHE_TIMEOUT)
+        return Response(status=HTTP_202_ACCEPTED)
+
+
+class AuthConfirm(UserAuthBaseView):
+
+    def post(self, request, format=None):
+        serializer = AuthConfirmSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        gateway_request_id = self.gateway_auth_confirm(serializer.data)
+        response_data = poll_for_data_in_cache(gateway_request_id)
+        return self.generate_response_from_callback(response_data)
+
+    def gateway_auth_confirm(self, request_data):
+        payload = ABDMRequestHelper.common_request_data()
+        payload['transactionId'] = request_data.pop('transactionId')
+        payload['credential'] = request_data['credential']
+        ABDMRequestHelper().gateway_post(UserAuthGatewayAPIPath.AUTH_CONFIRM, payload)
+        return payload['requestId']
+
+
+class GatewayAuthOnConfirm(UserAuthGatewayBaseView):
+
+    def post(self, request, format=None):
+        serializer = GatewayAuthOnConfirmSerializer(data=request.data)
```

Comment:
Except the serializer class everything here is ditto to `GatewayAuthOnInit`, may be better to DRY?

Reply:
Addressed in c3b6829c96f8af7fd2c0b1b62d30280469c7ea0d